chain/evm/deploy_contract.go

```diff
@@ -106,17 +124,137 @@ func deployContract(
 			})
 		}
 
+		var contractABI abi.ABI
+		contractABI, err = abi.JSON(strings.NewReader(rawABI))
+		if err != nil {
+			logger.WithError(err).Error("failed to parse raw JSON")
+			return
+		}
+
 		// hack begins here:
 		// constructor input arguments are already properly encoded, but
 		// we need to unpack them here because bind.DeployContract function
 		// expects arguments to come in "go" form
-		constructorArgs, err := contractAbi.Constructor.Inputs.Unpack(constructorInput)
+		constructorArgs, err := contractABI.Constructor.Inputs.Unpack(constructorInput)
 		whoops.Assert(err)
 		// hack ends here
 
 		logger.Info("deploying contract")
 
 		contractAddr, tx, _, err = bind.DeployContract(
+			txOpts,
+			contractABI,
+			bytecode,
+			ethClient,
+			constructorArgs...,
+		)
+		constructorArgs, _ = contractABI.Constructor.Inputs.Unpack(constructorInput)
+
+		whoops.Assert(err)
+		if tx.Type() == 2 {
+			logger.WithFields(log.Fields{
+				"tx-hash":          tx.Hash(),
+				"tx-gas-limit":     tx.Gas(),
+				"tx-gas-max-price": tx.GasFeeCap(),
+				"tx-gas-max-tip":   tx.GasTipCap(),
+				"tx-cost":          tx.Cost(),
+				"tx-type":          tx.Type(),
+			}).Info("eip-1559 tx executed")
+		} else {
+			logger.WithFields(log.Fields{
+				"tx-hash":      tx.Hash(),
+				"tx-gas-limit": tx.Gas(),
+				"tx-gas-price": tx.GasPrice(),
+				"tx-cost":      tx.Cost(),
+				"tx-type":      tx.Type(),
+			}).Info("legacy tx executed")
+		}
+	})
+	return
+}
+
+func deployContractArbitrum(
+	ctx context.Context,
+	ethClient arbbind.ContractBackend,
+	ks *arbkeystore.KeyStore,
+	signingAddr arbcommon.Address,
+	chainID *big.Int,
+	contractAbi arbabi.ABI,
+	bytecode []byte,
+	constructorInput []byte,
+	gasAdjustment float64,
+	txType uint8,
+) (contractAddr arbcommon.Address, tx *arbtypes.Transaction, err error) {
```

Comment:
This is basically a carbon copy for `deployContract`, rewritten to use types from the Arbitrum `go-ethereum` fork. I would have liked to use a more abstract definition to avoid code duplication, but we'll need a full abstraction package over address, hash, transaction types for non EVM support anyway - and as they say, premature optimisation is the root of all evil.